Add Documentation for IS Plugin Generator (#34)
ni/instrumentstudio-plugins: labview/docs/InstrumentStudio Plug-In Generator Guide.md

# InstrumentStudio Plug-In Generator for LabVIEW

- [InstrumentStudio Plug-In Generator for LabVIEW](#instrumentstudio-plug-in-generator-for-labview)
  - [Introduction](#introduction)
  - [Workflow](#workflow)
  - [Software support](#software-support)
  - [Installation](#installation)
  - [Developing a LabVIEW InstrumentStudio Plug-In](#developing-a-labview-instrumentstudio-plug-in)
  - [Deploying a LabVIEW InstrumentStudio plug-in](#deploying-a-labview-instrumentstudio-plug-in)
  - [Using a LabVIEW InstrumentStudio plug-in in InstrumentStudio](#using-a-labview-instrumentstudio-plug-in-in-instrumentstudio)
  - [Creating a package or installer to deploy a LabVIEW InstrumentStudio plug-in](#creating-a-package-or-installer-to-deploy-a-labview-instrumentstudio-plug-in)
  - [Note](#note)

---

## Introduction

> The InstrumentStudio Plug-In Generator tool allows users to create new plug-ins from a template by
> specifying a desired name and group. This tool provides a convenient starting point for developing

Reviewer: ```suggestion
The InstrumentStudio Plug-In Generator tool allows users to create new plug-ins with
specified name and group. This tool provides a convenient starting point for developing
```

any LabVIEW application as an InstrumentStudio plug-in.

---

## Workflow

![User Workflow](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/Workflow.png)

---

## Software support

NI Packages Dependency | Version Required
--- | ---
[LabVIEW (64-bit)](https://www.ni.com/en/support/downloads/software-products/download.labview.html#443865) | 2021 SP1 or later
[InstrumentStudio](https://www.ni.com/en/support/downloads/software-products/download.instrumentstudio.html#460631) | 2022 Q3 or later
[JKI VI Package Manager](https://www.ni.com/en/support/downloads/tools-network/download.jki-vi-package-manager.html#443251) | 2021 SP1 or later

---

## Installation

- Download the `InstrumentStudio Plug-In SDK` and `InstrumentStudio Plug-In Generator` packages from
  the latest release assets.

- Install the `InstrumentStudio Plug-In SDK` package, followed by the `InstrumentStudio Plug-In

>   Generator` package.

Reviewer: Have one point to download sdk followed by gen pkg, and add note emphasizing gen to installed first

## Developing a LabVIEW InstrumentStudio Plug-In

1. Create and save a new LabVIEW project.
2. From the project window, go to `Tools` → `Plug-In SDKs` → `InstrumentStudio Plug-In` → `Create
   InstrumentStudio Plug-in...`.
   - In the dialog, enter the `InstrumentStudio Plug-in Name` and `InstrumentStudio Plug-in Group`,

>      then click `Create InstrumentStudio Plug-in`.

Reviewer: This is a guide, so we can expand a little bit (only if necessary) at times

1. 'Create InstrumentStudio Plug-om' dialog box opens
2. Active project name will be auto-populated (give brief if needed)
3. Enter Plugin Name and Group
4. Click

      ![Create InstrumentStudio Plug-In](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/Create%20InstrumentStudio%20Plug-In.png)

>    - This will create a new InstrumentStudio plug-in library and a PPL build specification for the

Reviewer: I would prefer numbers over symbols

     plug-in in the project.

      ![InstrumentStudio Plug-In Library](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/InstrumentStudio%20Plug-In%20Library.png)

3. Add your logic/implementation to the top-level plug-in VI. By default, the VI shows off some of
   the basic features of LabVIEW InstrumentStudio plug-ins.

   ![Top-level plug-in VI Block diagram](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/Top-level%20VI%20Block%20diagram.png)

Note: For more information about the GPluginData file format and InstrumentStudio plug-in
development, please refer to the
[GPluginData-File-Format](https://github.com/ni/instrumentstudio-plugins/blob/main/labview/docs/GPluginData-File-Format.pdf) and
[G Plugin Development Guide](https://github.com/ni/instrumentstudio-plugins/blob/main/labview/docs/G%20Plugin%20Development%20Guide.pdf)
documents.

---

## Deploying a LabVIEW InstrumentStudio plug-in

> 1. Open the LabVIEW project which contains the InstrumentStudio plug-in.

Reviewer: ```suggestion
1. Open the Active LabVIEW project where the InstrumentStudio plug-in was created.
```

2. The generated plug-in comes with a Packed Project Library (PPL) build specification.

> 3. Build the PPL.

Reviewer: Explain - Right click-> Build

4. Copy the built plug-in files into the InstrumentStudio `Addons` directory, which is `C:\Program
   Files\National Instruments\InstrumentStudio\Addons` by default.
   - You may optionally install the plug-in files into a subdirectory of the `Addons` directory so
     its contents do not conflict with other plug-ins.

Note: For information about the recommended PPL build settings for an InstrumentStudio plug-in,
please refer to the 'Building and Deploying Release Plug-Ins' section in the
[G Plugin Development Guide](https://github.com/ni/instrumentstudio-plugins/blob/main/labview/docs/G%20Plugin%20Development%20Guide.pdf)
document.

---

## Using a LabVIEW InstrumentStudio plug-in in InstrumentStudio

1. Open InstrumentStudio and click `Manual Layout` to open the Edit Layout dialog.

> 2. In the dialog, the InstrumentStudio plug-in will be listed under the Add-Ons category within the
>    group specified in the `.gplugindata` file.

Reviewer: ```suggestion
2. The InstrumentStudio plug-in will be listed under the Add-Ons category within the
   group specified in the `.gplugindata` file.
```

>    group specified in the `.gplugindata` file.

Reviewer: Group specified in gplugindata file? or group mentioned in during creation? Because this is the first place where .gplugindata file is mentioned about, and there's no preview given for it

3. Choose the desired plug-in and create a large panel.
  
   ![Edit Layout](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/InstrumentStudio%20Edit%20Layout.png)

4. The layout will be populated with the plug-in UI as shown below.

   ![Plug-In SFP](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/InstrumentStudio%20Plug-In%20Soft%20Panel.png)

5. Modify the inputs in the UI to interact with the application.

---

## Creating a package or installer to deploy a LabVIEW InstrumentStudio plug-in

1. Open the LabVIEW project which contains the InstrumentStudio plug-in.
2. Create a Package or Installer build specification in the project. Refer to
   [Creating Build Specifications](https://www.ni.com/docs/en-US/bundle/labview/page/creating-build-specifications.html)
   for more information.

   ![Package or Installer build spec creation](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/Package%20or%20Installer%20build%20spec.png)

> 3. Open the `Destinations` page in the package or installer build specification properties.

Reviewer: Open ppl properties, navigate to Destination category

4. Set the destination directory to `C:\Program Files\National
   Instruments\InstrumentStudio\Addons\<Sub-directory name>`.

   ![Destination directory](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/Package%20or%20Installer%20Destination%20directory.png)

5. In the `Source Files` page, add the `PPL` build specification of the plug-in to the destination
   view. This action will automatically incorporate the files generated by the PPL into the package

>    or installer during the build process.

Reviewer: Second point as sub-point

  - [Creating a package or installer to deploy a LabVIEW InstrumentStudio plug-in](#creating-a-package-or-installer-to-deploy-a-labview-instrumentstudio-plug-in)
  - [Note](#note)

---

## Introduction

The InstrumentStudio Plug-In Generator tool allows users to create new plug-ins with specified name
and group. This tool provides a convenient starting point for developing any LabVIEW application as
an InstrumentStudio plug-in.

---

## Workflow

![User Workflow](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/Workflow.png)

---

## Software support

NI Packages Dependency | Version Required
--- | ---
[LabVIEW (64-bit)](https://www.ni.com/en/support/downloads/software-products/download.labview.html#443865) | 2021 SP1 or later
[InstrumentStudio](https://www.ni.com/en/support/downloads/software-products/download.instrumentstudio.html#460631) | 2022 Q3 or later
[JKI VI Package Manager](https://www.ni.com/en/support/downloads/tools-network/download.jki-vi-package-manager.html#443251) | 2021 SP1 or later

---

## Installation

Download the `InstrumentStudio Plug-In SDK` and `InstrumentStudio Plug-In Generator` packages from
the latest release assets.

> Note: Please install the `InstrumentStudio Plug-In SDK` package before installing `InstrumentStudio
> Plug-In Generator` package.

Reviewer: ```suggestion
> **Note:** Please install the `InstrumentStudio Plug-In SDK` package before installing `InstrumentStudio
Plug-In Generator` package.
```

Apply to all notes

## Developing a LabVIEW InstrumentStudio Plug-In

1. Create and save a new LabVIEW project.
2. From the project window, go to `Tools` → `Plug-In SDKs` → `InstrumentStudio Plug-In` → `Create
   InstrumentStudio Plug-in...`.
   1. This will open the `Create InstrumentStudio Plug-in` dialog.
   2. The name of the Active LabVIEW Project will be automatically populated in the `Active Project
      Name` indicator.  
      Note: The new InstrumentStudio plug-in will be created in this Active LabVIEW project.
   3. Enter the `InstrumentStudio Plug-in Name` and `InstrumentStudio Plug-in Group`.
   4. Click on the `Create InstrumentStudio Plug-in` button.

      ![Create InstrumentStudio Plug-In](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/Create%20InstrumentStudio%20Plug-In.png)

   5. This will create a new InstrumentStudio plug-in library and a PPL build specification for the
     plug-in in the Active LabVIEW project.

      ![InstrumentStudio Plug-In Library](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/InstrumentStudio%20Plug-In%20Library.png)

> 3. Add your logic/implementation to the top-level plug-in VI. By default, the VI shows off some of

Reviewer: Also, we can mention referring to examples in this repo for logic implementation

> 3. Add your logic/implementation to the top-level plug-in VI. By default, the VI shows off some of
>    the basic features of LabVIEW InstrumentStudio plug-ins.

Reviewer: ```suggestion
3. Default template is added to the library VIs, which contains some of the basic features of the LabVIEW InstrumentStudio plug-ins.
4. Add your logic/implementation to the top-level plug-in VI. 
```

5. Modify the inputs in the UI to interact with the application.

---

## Creating a package or installer to deploy a LabVIEW InstrumentStudio plug-in

1. Open the LabVIEW project which contains the InstrumentStudio plug-in.
2. Create a Package or Installer build specification in the project. Refer to
   [Creating Build Specifications](https://www.ni.com/docs/en-US/bundle/labview/page/creating-build-specifications.html)
   for more information.

   ![Package or Installer build spec creation](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/Package%20or%20Installer%20build%20spec.png)

3. Open the package or installer build specification properties, navigate to the `Destinations` page
   in the properties.
4. Set the destination directory to `C:\Program Files\National
   Instruments\InstrumentStudio\Addons\<Sub-directory name>`.

   ![Destination directory](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/Package%20or%20Installer%20Destination%20directory.png)

5. In the `Source Files` page, add the `PPL` build specification of the plug-in to the destination
   view.

>    - This action will automatically incorporate the files generated by the PPL into the package or

Reviewer: 'include' sounds simpler

# InstrumentStudio Plug-In Generator for LabVIEW

- [InstrumentStudio Plug-In Generator for LabVIEW](#instrumentstudio-plug-in-generator-for-labview)
  - [Introduction](#introduction)
  - [Workflow](#workflow)
  - [Software support](#software-support)
  - [Installation](#installation)

>   - [Developing a LabVIEW InstrumentStudio Plug-In](#developing-a-labview-instrumentstudio-plug-in)

Reviewer 2: This one is caps 'Plug-In'. Most others are 'plug-in' in the titles here. Ensure consistency.

   features of the LabVIEW InstrumentStudio plug-ins.

   ![Top-level plug-in VI Block diagram](./images/InstrumentStudio%20Plug-In%20Generator%20Guide/Top-level%20VI%20Block%20diagram.png)

4. Add your logic implementation to the top-level plug-in VI.
      - For reference, please refer to the example plugins in the repository.

**Note:** For more information about the GPluginData file format and InstrumentStudio plug-in
development, please refer to the
[GPluginData-File-Format](https://github.com/ni/instrumentstudio-plugins/blob/main/labview/docs/GPluginData-File-Format.pdf) and
[G Plugin Development Guide](https://github.com/ni/instrumentstudio-plugins/blob/main/labview/docs/G%20Plugin%20Development%20Guide.pdf)
documents.

---

## Deploying a LabVIEW InstrumentStudio plug-in

1. Open the Active LabVIEW project where the InstrumentStudio plug-in was created.
2. The generated plug-in comes with a Packed Project Library (PPL) build specification.
3. To build the PPL, right-click on the desired build specification and choose 'Build'.
4. Copy the built plug-in files into the InstrumentStudio `Addons` directory, which is `C:\Program

>    Files\National Instruments\InstrumentStudio\Addons` by default.

Reviewer 3: You may also want to mention plugins can install to a user defined directory configured in IS preferences dialog for plugins. This directory is intended to facilitate installation of plugins without requiring administrative permissions. Configuring this preference requires IS to be launched with administrator privileges.

![image](https://github.com/user-attachments/assets/1bd0b3b9-9bc9-4800-a83b-c356e65a9ac6)

   - You may optionally install the plug-in files into a subdirectory of the `Addons` directory so
     its contents do not conflict with other plug-ins.

**Note:** For information about the recommended PPL build settings for an InstrumentStudio plug-in,
please refer to the 'Building and Deploying Release Plug-Ins' section in the
[G Plugin Development Guide](https://github.com/ni/instrumentstudio-plugins/blob/main/labview/docs/G%20Plugin%20Development%20Guide.pdf)
document.

---

## Using a LabVIEW InstrumentStudio plug-in in InstrumentStudio

1. Open InstrumentStudio and click `Manual Layout` to open the Edit Layout dialog.
2. The InstrumentStudio plug-in will be listed under the Add-Ons category within the group specified
   at the time of the plug-in creation.  

>    2.1  If `InstrumentStudio Plug-in Group` is not specified during plug-in creation, the plug-in

Reviewer 4: These are inconsistent with `1. a.` above. I'm not sure Markdown even recognizes `2.1` as a list item.

Author: Thank you for pointing this out. I have updated the bulletins in the document to ensure consistency.

   1. This will open the `Create InstrumentStudio Plug-in` dialog.
   2. The name of the Active LabVIEW Project will be automatically populated in the `Active Project
      Name` indicator.  

>       **Note:** The new InstrumentStudio plug-in will be created in this active LabVIEW project.

Reviewer 4: GitHub Markdown has a special extension for notes and other alerts: https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#alerts

Author: Thank you for sharing the information about the special extension for notes and other alerts. I have updated the 'Notes' in the document to follow the alert syntax wherever applicable.
